mergekit/scripts/extract_lora.py

```diff
@@ -0,0 +1,194 @@
+
+    return A, B
+
+
+def find_all_linear_names(model):
+    cls = (bnb.nn.Linear4bit, bnb.nn.Linear8bitLt, torch.nn.Linear, QuantLinear)
+
+    names = []
+    for name, module in model.named_modules():
+        if (
+            isinstance(module, cls)
+            or "Linear" in module.__class__.__name__
+            and module.__class__.__name__ not in ("LlamaLinearScalingRotaryEmbedding",)
+        ):
+            names.append(name)
+
+    return names
+
+
+def get_linear_module_names(model_id):
+    model = AutoModelForCausalLM.from_pretrained(
+        model_id, state_dict={}, device_map="meta"
+    )  # avoid loading weights as we won't need them
+    linear_module_names = find_all_linear_names(model)
+
+    return linear_module_names
+
+
+def create_peft_config(base_model_name_or_path, rank, alpha, target_modules):
+    return {
+        "alpha_pattern": {},
+        "auto_mapping": None,
+        "base_model_name_or_path": base_model_name_or_path,
+        "bias": "none",
+        "fan_in_fan_out": False,
+        "inference_mode": True,
+        "init_lora_weights": True,
+        "layers_pattern": None,
+        "layers_to_transform": None,
+        "loftq_config": {},
+        "lora_alpha": alpha,
+        "lora_dropout": 0,
+        "megatron_config": None,
+        "megatron_core": "megatron.core",
+        "modules_to_save": None,
+        "peft_type": "LORA",
+        "r": rank,
+        "rank_pattern": {},
+        "revision": None,
+        "target_modules": target_modules,
+        "task_type": "CAUSAL_LM",
+        "use_rslora": False,
+    }
+
+
+@click.command("mergekit-extract-lora")
+@click.argument("out_path", type=click.Path())
+@click.option("--base-model", type=str, help="Model ID or path to use as base model")
+@click.option(
+    "--finetuned-model",
+    type=str,
+    help="Model ID or path to use as PEFT extraction target model",
+)
+@click.option(
+    "--rank", type=int, default=32, help="Rank for the low-rank decomposition"
+)
+@click.option(
+    "--device",
+    type=str,
+    default=None,
+    help="PyTorch device to perform SVD computation on",
+)
+def main(out_path: str, base_model: str, finetuned_model: str, rank: int, device: str):
+    """
+    Decomposes delta weights between a base model and a finetuned model and saves a PEFT model.
+    """
+    os.makedirs(out_path, exist_ok=True)
+
+    base_model_ref = ModelReference.parse(base_model)
+    finetuned_model_ref = ModelReference.parse(finetuned_model)
+
+    base_model_config = AutoConfig.from_pretrained(base_model_ref.model.path)
+
+    linear_module_names = get_linear_module_names(base_model_ref.model.path)
+    finetuned_model_linear_module_names = get_linear_module_names(
+        finetuned_model_ref.model.path
+    )
+
+    assert set(linear_module_names) == set(
+        finetuned_model_linear_module_names
+    ), "Model architecture mismatch"
+
+    base_loader = LazyTensorLoader(base_model_ref.tensor_index(), lazy_unpickle=True)
```

Comment:
It'd be good to add an option for `--lazy-unpickle/--no-lazy-unpickle` for the occasional weird file that the lazy unpickler can't handle.